### ISSUE TITLE
drrun -stat option does not display any stats

### ISSUE THREAD

**Issue by nirhasabnis** [DynamoRIO/dynamorio#1948]
Hi,

I am trying to get total CPU time of app execution using DynamoRio, and I found -stat/-mem option for drrun tool. But I used this option and I did not get any stats output. All I get is the output from the app.
I checked the source code and I found that these options do not do anything on Linux? Am I missing something here?

Example:
$ bin64/drrun -stats -- uname
Linux

I do not see any stats information in above command.

My setup is:
Ubuntu-14.04 x86_64
drrun version 6.1.1 -- build 3

Thanks,

**Comment by derekbruening** [comment]
Right, -stats and -mem only exist for Windows, and drrun should not show them in the usage message on *nix.  This issue can serve as a feature request for implementing them, which would require switching (only for these options: we want to directly exec by default) to a fork model which drinjectlib (at last partially, for ptrace injection) supports for *nix.

Xref runstats.c which contains all of the code we would need.  In fact for now you could use this from a build dir (not an install dir b/c we don't export runstats):

```
% bin64/drrun -- bin64/runstats -mem uname
```

We would welcome this as a contribution.

**Comment by nirhasabnis** [comment]
Thanks.

I could run `bin64/drrun -- bin64/runstats uname` command and get the timing info. But I found that the numbers match with those returned by ``/usr/bin/time bin64/drrun -- uname``. This makes me believe that runstats report execution time of drrun. What I mean is that, as per my understanding, there are two components when drrun executes a program: (1) DynamoRio's setup (startup), and (2) execution time of the program itself. I was under the impression that runstat returns only (2), but it seems that it returns (1)+(2). Can you please clarify? In case, it returns (1)+(2), is there a way to measure only (2)? Thanks.

**Comment by derekbruening** [comment]
> I could run bin64/drrun -- bin64/runstats uname command and get the timing info. But I found that the 
> numbers match with those returned by /usr/bin/time bin64/drrun -- uname. This makes me believe that > runstats report execution time of drrun.

?  runstats has no relation to drrun and does not know of its existence in the command line "bin64/drrun -- bin64/runstats uname".  In the command line "/usr/bin/time bin64/drrun -- uname", time will measure its child process, which starts as drrun but then execs into DR running uname.

> What I mean is that, as per my understanding, there are two components when drrun executes a 
> program: (1) DynamoRio's setup (startup), and (2) execution time of the program itself.

drrun does an exec of DR who then loads the app and runs the app under its control in its code cache.  Certainly there is some initialization prior to running the app code in the cache, but there are also overhead and actions taken throughout the run, so it doesn't really make sense to talk about it as though there's some strict setup phase and then some native execution that has no impact from DR.

**Comment by nirhasabnis** [comment]
Right, I understand both DR's initialization overhead and the overhead of app code throughout the run. What I am looking for is a way to measure the overhead of these two components separately. I want to get an estimation of how much of the overhead comes from one component vs the other. 

I know that DR's overhead on SPECCPU 2006 is 12%. I wanted to know how much of 12% comes from one vs the other. If there already exists such documentation somewhere, can you please point me to it?

**Comment by derekbruening** [comment]
> I know that DR's overhead on SPECCPU 2006 is 12%. I wanted to know how much of 12% comes 
> from one vs the other. If there already exists such documentation somewhere, can you please point me 
> to it?

I'm not sure why you are asking about initialization.  ￼The time spent in initial process initialization for SPECCPU is essentially zero.  Initialization just doesn't matter for pretty much any app.  See my thesis section 7.2 which has several tables of data.  Probably what you really mean is time to bring new code into the code cache, which happens throughout execution, and not the process startup code, but that is also negligible for SPECCPU: only for gcc with its separate, short runs is it over 1%.
